File: server/ops/environment/sampler.js
// Real system-load sampling — CPU, memory, and Jarvis's own process memory.
// Deliberately does NOT use os.loadavg(): it returns [0, 0, 0] on Windows
// unconditionally (a documented Node/libuv limitation, not a bug this
// project could work around), which would make every reading on this
// project's own target platform silently useless. Real CPU usage instead
// comes from sampling os.cpus() TWICE and comparing the delta between idle
// and total time across the interval — a single os.cpus() snapshot has no
// usable percentage in it at all (it's a cumulative counter since boot).

import os from 'node:os';
import { recordSample, pruneOld } from './env-store.js';

const SAMPLE_INTERVAL_MS = 30 * 1000;
const PRUNE_EVERY_N_SAMPLES = 200; // roughly once per ~100 minutes at the interval above

function cpuTimesSnapshot() {
  return os.cpus().map((c) => ({ idle: c.times.idle, total: Object.values(c.times).reduce((a, b) => a + b, 0) }));
}

/** Real CPU busy percentage over the interval between `prev` and a fresh snapshot — null if the core count changed (a real, if rare, case: e.g. Windows power-plan core parking) rather than a misleading number. */
function cpuPercentSince(prev) {
  const now = cpuTimesSnapshot();
  if (!prev || prev.length !== now.length) return { pct: null, snapshot: now };
  let idleDelta = 0;
  let totalDelta = 0;
  for (let i = 0; i < now.length; i++) {
    idleDelta += now[i].idle - prev[i].idle;
    totalDelta += now[i].total - prev[i].total;
  }
  if (totalDelta <= 0) return { pct: null, snapshot: now };
  const pct = Math.max(0, Math.min(100, (1 - idleDelta / totalDelta) * 100));
  return { pct, snapshot: now };
}

let lastCpuSnapshot = null;
let sampleCount = 0;
let timer = null;

function takeSample() {
  const { pct: cpuPct, snapshot } = cpuPercentSince(lastCpuSnapshot);
  lastCpuSnapshot = snapshot;

  const freeBytes = os.freemem();
  const totalBytes = os.totalmem();
  const memFreePct = totalBytes > 0 ? (freeBytes / totalBytes) * 100 : null;

  const rssBytes = process.memoryUsage().rss;

  recordSample({ cpuPct, memFreePct, rssBytes });

  sampleCount++;
  if (sampleCount % PRUNE_EVERY_N_SAMPLES === 0) {
    try {
      pruneOld();
    } catch (err) {
      console.error('[ops] env_samples prune failed:', err);
    }
  }
}

/** Starts periodic sampling — called once from server.js. Idempotent. The very first sample has cpuPct:null (no prior snapshot to diff against yet) — genuinely unknown, not zero. */
export function startSampling() {
  if (timer) return;
  takeSample();
  timer = setInterval(takeSample, SAMPLE_INTERVAL_MS);
  timer.unref?.();
}

/** A fresh reading right now, bypassing the sample interval — check_environment.js's own "how loaded are things RIGHT NOW" path. Does not itself write a row (avoids double-counting against the regular cadence) unless `record:true` is passed. */
export function sampleNow({ record = false } = {}) {
  const { pct: cpuPct, snapshot } = cpuPercentSince(lastCpuSnapshot);
  if (record) lastCpuSnapshot = snapshot;
  const freeBytes = os.freemem();
  const totalBytes = os.totalmem();
  const memFreePct = totalBytes > 0 ? (freeBytes / totalBytes) * 100 : null;
  const rssBytes = process.memoryUsage().rss;
  const reading = { cpuPct, memFreePct, rssBytes, freeBytes, totalBytes };
  if (record) recordSample(reading);
  return reading;
}

File: backend/jarvis/ops/environment/sampler.py
"""Taking a real reading of CPU, memory and this process's own footprint.

psutil rather than a hand-rolled read: the platform-specific one this replaces
had to diff two `os.cpus()` snapshots because the obvious call returns zeros on
Windows, which is the platform this actually ships on. One dependency removes an
entire class of "correct on the machine it was written on" bug.

**The first CPU reading of a process is genuinely unknown, not zero.** A
percentage is a measurement over an interval, and there is no interval yet. It
is reported as None, and the baseline ignores it, rather than being recorded as
an idle machine that never was.
"""

from __future__ import annotations

import logging
import os
import threading
from typing import Any

import psutil

from . import store

logger = logging.getLogger(__name__)

SAMPLE_INTERVAL_S = 30.0
PRUNE_EVERY_N = 200

ENABLE_ENV = "JARVIS_ENV_SAMPLER"

_timer: threading.Timer | None = None
_samples_taken = 0
_primed = False


def _cpu_percent() -> float | None:
    """Busy percentage since the previous call. None on the very first one."""
    global _primed
    value = psutil.cpu_percent(interval=None)
    if not _primed:
        _primed = True
        return None                      # no interval to have measured over yet
    return float(value)


def read_now() -> dict[str, Any]:
    """A reading, without recording it — the pull path's "how loaded is it right
    now". Recording here would double-count against the regular cadence."""
    memory = psutil.virtual_memory()
    return {
        "cpuPct": _cpu_percent(),
        "memFreePct": float(memory.available) / float(memory.total) * 100 if memory.total else None,
        "rssBytes": int(psutil.Process().memory_info().rss),
        "freeBytes": int(memory.available),
        "totalBytes": int(memory.total),
        "cpuCount": psutil.cpu_count(logical=True),
    }


def take_sample() -> dict[str, Any]:
    global _samples_taken
    reading = read_now()
    store.record_sample(cpu_pct=reading["cpuPct"], mem_free_pct=reading["memFreePct"],
                        rss_bytes=reading["rssBytes"])
    _samples_taken += 1
    if _samples_taken % PRUNE_EVERY_N == 0:
        try:
            store.prune_old()
        except Exception:  # noqa: BLE001 — housekeeping must not break sampling
            logger.exception("pruning env samples failed")
    return reading


def is_enabled() -> bool:
    return os.environ.get(ENABLE_ENV) == "1"


def start() -> bool:
    global _timer
    if not is_enabled() or _timer is not None:
        return False

    def run() -> None:
        global _timer
        try:
            take_sample()
        except Exception:  # noqa: BLE001
            logger.exception("taking a system sample failed")
        _timer = threading.Timer(SAMPLE_INTERVAL_S, run)
        _timer.daemon = True
        _timer.start()

    run()
    return True


def stop() -> None:
    global _timer
    if _timer is not None:
        _timer.cancel()
        _timer = None


def reset_for_tests() -> None:
    global _samples_taken, _primed
    stop()
    _samples_taken = 0
    _primed = False
